v1.1/performance-best-practices-overview.md

```diff
@@ -0,0 +1,196 @@
+
+### Use `INSERT` with the `RETURNING` Clause to Generate Unique IDs
+
+If, for some reason, you cannot use `UUID` to generate the unique IDs, then you might resort to the common practice to generate unique IDs is to use roundtrip `SELECT` in a transaction. Instead, for improved performance, [use the `RETURNING` clause with the `INSERT` statement](insert.html#insert-and-return-values) instead.
+
+#### Generate Monotonically-Increasing Unique IDs
+
+Suppose the table schema is as follows:
+
+~~~ sql
+> CREATE TABLE X (
+ID1 INT,
+ID2 INT,
+ID3 INT DEFAULT 1,
+PRIMARY KEY (ID1,ID2));
+~~~
+
+The common approach to generate monotonically-increasing Unique IDs is a transaction using the `SELECT` statement:
+
+~~~ sql
+> BEGIN;
+
+> INSERT INTO X VALUES (1,1,1)
+
+> ON CONFLICT (ID1,ID2)
+
+> DO UPDATE SET ID3=X.ID3+1;
+
+> SELECT * FROM X WHERE ID1=1 AND ID2=1;
+
+> COMMIT;
+~~~
+
+However, the performance best practice is to use a `RETURNING` clause with `INSERT` instead of the transaction:
+
+~~~ sql
+> INSERT INTO X VALUES (1,1,1),(2,2,2),(3,3,3)
+ON CONFLICT (ID1,ID2)
+DO UPDATE SET ID3=X.ID3 + 1
+RETURNING ID1,ID2,ID3;
+~~~
+
+#### Generate Random Unique IDs
+
+Suppose the table schema is as follows:
+
+~~~ sql
+> CREATE TABLE X (
+ID1 INT,
+ID2 INT,
+ID3 SERIAL,
+PRIMARY KEY (ID1,ID2));
+~~~
+
+The common approach to generate random Unique IDs is a transaction using the `SELECT` statement:
+
+~~~ sql
+> BEGIN;
+
+> INSERT INTO X VALUES (1,1);
+
+> SELECT * FROM X WHERE ID1=1 AND ID2=1;
+
+> COMMIT;
+~~~
+
+However, the performance best practice is to use a `RETURNING` clause with `INSERT` instead of the transaction:
+
+~~~ sql
+> INSERT INTO X VALUES (1,1),(2,2),(3,3)
+RETURNING ID1,ID2,ID3;
+~~~
+
+## Indexes Best Practices
+
+### Use Secondary Indexes
+
+You can improve the performance of queries using columns besides the primary key with secondary indexes. You can create them:
+
+- At the same time as the table with the `INDEX` clause of [`CREATE TABLE`](create-table.html#create-a-table-with-secondary-indexes). In addition to explicitly defined indexes, CockroachDB automatically creates secondary indexes for columns with the [Unique constraint](unique.html).
+- For existing tables with [`CREATE INDEX`](create-index.html).
+- By applying the Unique constraint to columns with [`ALTER TABLE`](alter-table.html), which automatically creates an index of the constrained columns.
+
+To create the most useful secondary indexes, check out our [best practices](indexes.html#best-practices).
+
+### Create Indexes before Performing `JOIN` Operation
+
+CockroachDB supports both [merge joins](https://en.wikipedia.org/wiki/Sort-merge_join) and [hash joins](https://en.wikipedia.org/wiki/Hash_join). While performing the [`JOIN`](table-expressions.html#join-expressions) operation, CockroachDB uses the merge join instead of the hash join whenever possible, because merge joins are computationally and memory-wise better performant than hash joins. 
```

Comment:
I think we should combine the "why create indexes" content with the intro, given that the best practice is to create indexes:

> CockroachDB uses merge joins whenever possible because they are more performant than hash joins computationally and in terms of memory. However, merge joins are possible only when the tables being joined are indexed on the relevant columns; when this condition is not met, CockroachDB resorts to the slower hash joins.